### ISSUE TITLE
Mask Invalid Token Logits during Inference for Consistency with MSL

### ISSUE THREAD

**Issue by yuyq18** [WANGBohaO-jpg/MSL#5]
Thank you for the great work on Masked Softmax Loss (MSL) and for releasing the code.

While reading through the implementation, we noticed a **potential inconsistency between training and inference**.

During training, invalid tokens are masked out in the `compute_loss` function — their logits are excluded from the softmax denominator, effectively treated as `-inf`. This is key to aligning model optimization with the constrained generation space.

However, during inference (e.g., constrained beam search), the model `forward` function outputs **unmasked logits over the full vocabulary**. Invalid tokens are only filtered after the logits are computed, when selecting the next token.

We suggest applying the same masking logic to the logits in the model forward function, not just during loss computation. This ensures that both training and inference operate on the same constrained token space and improves alignment between optimization and decoding.

**Comment by WANGBohaO-jpg** [comment]
Thank you very much for your careful reading of our code and your valuable feedback!

Regarding your concern about the potential inconsistency between training and inference:  
In our inference pipeline, we are using constrained beam search as implemented in the HuggingFace Transformers library, specifically via the `prefix_allowed_tokens_fn` function. In this implementation, the logits of invalid tokens are indeed set to `-inf` **before** applying the softmax, effectively assigning zero probability to these tokens. This is the same masking logic as used during training in the `compute_loss` function, where invalid tokens are also excluded from the softmax denominator by setting their logits to `-inf`.

Therefore, from our understanding, both training and inference handle the constrained token space in a consistent manner—invalid tokens are masked out by setting their logits to `-inf` before the softmax operation.

If you have observed any specific inconsistency or if there is a particular scenario where the masking does not align as expected, could you please provide more details or point to a concrete example in the code? We would be happy to investigate further and clarify any potential issues.

**Comment by yuyq18** [comment]
The code in constrained beam search:
Stage | Mask position | Softmax denominator | Effect on valid-token probability
-- | -- | -- | --
Training (MSL) | mask firstfθ(z)=−∞ for invalid before softmax | ∑z∈𝒁valid exp fθ(z) | No dilution
Inference (code above) | softmax first → mask later | ∑z∈𝒁 exp fθ(z) | Valid-token mass is diluted by invalid logits

**Comment by yuyq18** [comment]
Thanks for the clarification!

However, there's still a mathematical inconsistency between training and inference.

The code in constrained beam search: 
```
next_token_logits = outputs.logits[:, -1, :]

next_token_logits = self.adjust_logits_during_generation(next_token_logits, cur_len=cur_len)

# 1️⃣ Softmax over the **full vocabulary** 𝒁
next_token_scores = nn.functional.log_softmax(
    next_token_logits, dim=-1
)  # (batch_size * num_beams, vocab_size)

# 2️⃣ Post-hoc masking of invalid tokens
next_token_scores_processed = logits_processor(input_ids, next_token_scores)

next_token_scores = next_token_scores_processed + beam_scores[:, None].expand_as(next_token_scores)
```



Stage | Mask position | Softmax denominator | Effect on valid-token probability
-- | -- | -- | --
Training (MSL) | mask first f_\theta(z) = -\infty for invalid before softmax | $\sum_{z \in \mathcal{Z}{\text{valid}}} \exp(f_\theta(z))$ | No dilution
Inference (code above) | softmax first → mask later | $\sum_{z \in \mathcal{Z}} \exp(f_\theta(z))$ | Valid-token mass is diluted by invalid logits


Mathematically:

```markdown
$$
P_{train}(z) = \frac{\exp(f_\theta(z))}{\sum_{z' \in \mathcal{Z}_{\mathrm{valid}}} \exp(f_\theta(z'))}
$$

$$
P_{infer}(z) =
\begin{cases}
\frac{\exp(f_\theta(z))}{\sum_{z' \in \mathcal{Z}} \exp(f_\theta(z'))}, & z \in \mathcal{Z}_{\mathrm{valid}}
0, & z \notin \mathcal{Z}_{\mathrm{valid}}
\end{cases}
$$
```

**Comment by WANGBohaO-jpg** [comment]
Thank you for your careful review and for pointing out this important inconsistency.

I rechecked the source code of Transfomers library. You are absolutely correct—`prefix_allowed_tokens_fn` in Transformers first applies softmax over the full vocabulary and then masks out invalid tokens. This is inconsistent with the training phase in MSL because it causes the sum of probabilities over valid tokens to be less than 1 at each step.

To fix this, I will modify the `inference.py` so that the logits of invalid tokens are set to `-inf` before applying softmax.

Thank you again for catching this. I will update the code accordingly.
